### ISSUE TITLE
Reimplement LogNamespace= with bpf

### ISSUE THREAD

**Issue by daandemeyer** [systemd/systemd#24589]
### Component

systemd-journald

### Is your feature request related to a problem? Please describe

LogNamespace= currently implies a mount namespace for services that use it. This comes with a few issues and is often unexpected. Ideally using LogNamespace= would be less intrusive

### Describe the solution you'd like

We can use seccomp notify (see https://people.kernel.org/brauner/the-seccomp-notifier-new-frontiers-in-unprivileged-container-development) to intercept connect() syscalls to /dev/log and /run/systemd/journal/socket and connect to their log namespace equivalents instead. This allows us to implement log namespaces without requiring the services using it to use a separate mount namespace.

From reading up on the seccomp library, the implementation might look as follows:

- We configure `SCMP_ACT_NOTIFY` for the connect() syscall if a log namespace is configured. 
- We retrieve the seccomp notify fd in the forked child process after applying the syscall filter and send it to pid1 via a unix socket
- We add the seccomp notify fd to the event loop
- When the fd is notified, we check if the intercepted syscall is a connect() syscall to /dev/log or /run/systemd/journal/socket and look up the exec context of the pid doing the syscall to see if it has a log namespace configured. If the connect() is not to one of these paths or the pid's associated unit does not have a log namespace configured, we allow the connect() to continue using `SECCOMP_USER_NOTIF_FLAG_CONTINUE`
- Otherwise, we retrieve the socket fd passed to connect() using pidfd_get_fd()
- We then connect that fd to the log namespace equivalent socket of /dev/log or /run/systemd/journal/socket, depending on which one was passed to connect()

cc @poettering does this make sense?

### Describe alternatives you've considered

BPF allows intercepting connect() per cgroup (see https://lwn.net/Articles/749326/) which would allow us to rewrite the sockaddr to point to the correct path. The advantage of this approach would be that all redirection happens in the kernel and pid1 is not involved at all. However, this bpf feature does not support unix sockets yet so it isn't an option until that is added.

### The systemd version you checked that didn't have the feature you are asking for

main

**Comment by poettering** [comment]
Hmm, what I don't like about this is that connect() invocations must be bumped off an external process then. I'd much rather see that the kernel would just allow us to rewrite the sockaddr inside ebpf. I mean, it allows that for IPv4 and IPv6, no? Sounds like adding this for AF_UNIX should not be a big political issue, but more completing the existing work?

**Comment by DaanDeMeyer** [comment]
Yeah, I looked into this more and emulating connect() is not going to be trivial, ebpf will be much simpler once it has support for unix sockets.

There might also be perf implications involved with bumping every connect() call off to some systemd daemon, bpf should be much faster in that regard.

**Comment by poettering** [comment]
So, the tricky part about rewriting `sockaddr_un` is that userspace can pass it in with very short sizes, which might not be enough space to rewrite it.

So ideally, ebpf would have an API to synthesize an entirely new sockaddr_un and pass that on, instead of just rewriting the existing one.

**Comment by DaanDeMeyer** [comment]
> So, the tricky part about rewriting sockaddr_un is that userspace can pass it in with very short sizes, which might not be enough space to rewrite it.

But `sockaddr_un` is defined as follows so we always have 108 bytes available for rewriting no? I'm not too familar with the details of sockaddrs

```
struct sockaddr_un {
	__kernel_sa_family_t sun_family; /* AF_UNIX */
	char sun_path[UNIX_PATH_MAX];	/* pathname */
};
```

**Comment by poettering** [comment]
When you pass it to syscalls you specify the size in a separate parameter, and you want to specify only the size you actually make use of. See our SOCKADDR_UN_LEN() macro that calculates the right size to pass to connect(), sendto(), bind() and friends

sockadrr_un filled in for /dev/log typically passes a size of 13 hence (4 for sun_family + 8 for the string + trailing NUL byte), and that's mighty little....

**Comment by mildred** [comment]
I believe this could fix an issue like this one and be very useful for containers

- https://github.com/containers/podman/issues/19177

Perhaps there are other solutions than using seccomp or namespaces for logging? Why can't an environment variable be used to tell the log namespace or file path to use when logging?

**Comment by daandemeyer** [comment]
> I believe this could fix an issue like this one and be very useful for containers
> 
> - https://github.com/containers/podman/issues/19177
> 
> Perhaps there are other solutions than using seccomp or namespaces for logging? Why can't an environment variable be used to tell the log namespace or file path to use when logging?

So environment variables aren't guaranteed to be inherited by child processes. But even worse than that, syslog() is hard coded to use /dev/log and nothing else.

Anyways there's ways forward using bpf which I hope to implement sooner rather than later.
